ASoC: Intel: Boards: tgl_max98373: Add BT offload support

yongzhi1 opened the pull request:
BT audio offload is internally configured with virtual GPIOs,
the port assignment is fixed on SSP2 for TGL/ADL platforms.

This feature is enabled by coreboot without ACPI property.

Signed-off-by: Yong Zhi <yong.zhi@intel.com>

yongzhi1 commented:
The feature will first appear on ADL, would like to receive early feedback for code change on this and sof_sdw.c as well, thanks!!

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -922,6 +952,7 @@ static const struct platform_device_id board_ids[] = {
 					SOF_SPEAKER_AMP_PRESENT |
 					SOF_MAX98373_SPEAKER_AMP_PRESENT |
 					SOF_RT5682_SSP_AMP(1) |
+					SOF_BT_OFFLOAD_ENABLED |
```
It might be wiser to use a macro that allows us to deal with other values, e.g. 
SOF_BT_OFFLOAD_SSP(2)

plbossart replied:
also I am confused, you were mentioning earlier this is an ADL feature but it's added here for TGL?

yongzhi1 replied:
We are using TGL for development vehicle at the moment.

yongzhi1 replied:
> It might be wiser to use a macro that allows us to deal with other values, e.g.
> SOF_BT_OFFLOAD_SSP(2)

Will give it a try, thanks!!

plbossart replied:
> We are using TGL for development vehicle at the moment.

There's actually no harm in defining more dailinks than what is used by the topology, so we could add this by default in TGL.
The other way around would be a mistake, we can't have a topology that refers to a non-existing dailink.

yongzhi1 replied:
That's great to know, will send 2nd version.

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -777,8 +785,27 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
+		links[id].id = id;
+		links[id].cpus = &cpus[id];
+		links[id].cpus->dai_name = "SSP2 Pin";
+		links[id].name = "SSP2-Codec";
```
SSP2-NoCodec?

@ranj063 IIRC the dai link name is used in the topology, right? It may be useful to tag this link as different from the rest?

ranj063 replied:
should this be hardcoded to 2? Why not set it in the driver_data just like the rt5682/speaker SSP ports?

ranj063 replied:
yes @plbossart this dai_link name is used in topology but not specifically for SSP2 today at least for this machine. Why not call it SSP2-BT to be more specific instead of NoCodec?

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -777,8 +785,27 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
+		id++;
 	}
 
+	/* CNVi bt-offload */
```
Not sure if CNVi means anything for upstream?

yongzhi1 replied:
Ack, will remove the Intel "jargon".

plbossart reviewed (changes requested):
looks mostly good but some suggested changes below.. Thansk @yongzhi1

yongzhi1 commented:
V2 update:

1. Added macros to avoid hard-coded ssp #
2. Changed "SSP2-Nocodec" to "SSP2-BT"
3. Removed cNVi in the patch subject

Since Google is also interested to have BT-offload enabled on TGL/delbin, it may as well make the changes for upstream.

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -780,6 +794,32 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
+		links[id].num_codecs = ARRAY_SIZE(dummy_component);
+		links[id].platforms = platform_component;
+		links[id].num_platforms = ARRAY_SIZE(platform_component);
+		links[id].nonatomic = true;
```
I don't think this is 'nonatomic' is required for BE links.

This seems to be a copy/paste mistake. i will add a PR to clean this up in all drivers.

yongzhi1 replied:
Ack, thanks for catching this.

plbossart reviewed (changes requested):
please remove the nonatomic part it's not needed

plbossart reviewed (approved):
Thank you @yongzhi1

kv2019i reviewed (approved):
Thanks @yongzhi1 !
